From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: HaHaWTH <102713261+HaHaWTH@users.noreply.github.com>
Date: Tue, 9 Nov 2077 00:00:00 +0800
Subject: [PATCH] Rewrite entity despawn time

HOW IT WORKS:
In the default implementation, totalEntityAge only increments when an entity is actively ticked.
Consequently, if an entity resides in a weak-loaded chunk, its aging process effectively pauses, allowing it to exist far beyond the configured despawn time.

This patch addresses this discrepancy by synchronizing entity age with absolute server time:
Introduces lastTickTime to record the MinecraftServer.currentTick when an entity was last ticked.
Once the entity is ticked again or unloaded, the system calculates the elapsed time (currentTick - lastTickTime), and adds it to the totalEntityAge.
Also, it moves the despawn logic of projectiles to their own tick method (see ThrowableProjectile#tick), making it faster to despawn massive stuck projectiles.

Benchmarks (60,000 snowballs):
Before: Around 120 seconds, server thread is struggling to process projectile physics.
After: 1.5 seconds

Brings the ability to despawn weak-loaded entities once they are ticked,
a solution for https://github.com/PaperMC/Paper/issues/12986

License: GPL-3.0-only (https://www.gnu.org/licenses/gpl-3.0.html)

Additional term under GNU GPLv3 section 7(b):

Any copy or modified version containing a substantial portion of the
original contributions in this patch must preserve the following
author attribution in the patch header, source code, or another
prominent legal notice accompanying the corresponding source:

Original patch by HaHaWTH <102713261+HaHaWTH@users.noreply.github.com>

diff --git a/ca/spottedleaf/moonrise/patches/chunk_system/level/entity/ChunkEntitySlices.java b/ca/spottedleaf/moonrise/patches/chunk_system/level/entity/ChunkEntitySlices.java
index 81e72a1cd2de1bfeb4c78999148dd2e95ed782c2..2b90941bf6824719531582c27e640514742f0dd2 100644
--- a/ca/spottedleaf/moonrise/patches/chunk_system/level/entity/ChunkEntitySlices.java
+++ b/ca/spottedleaf/moonrise/patches/chunk_system/level/entity/ChunkEntitySlices.java
@@ -183,6 +183,7 @@ public final class ChunkEntitySlices {
                 // removed by us below
                 continue;
             }
+            if (entity.detectDespawnTime(true)) continue; // Leaf - Rewrite entity despawn time
             if (entity.shouldBeSaved()) {
                 PlatformHooks.get().unloadEntity(entity);
                 if (entity.isVehicle()) {
diff --git a/ca/spottedleaf/moonrise/patches/chunk_system/level/entity/EntityLookup.java b/ca/spottedleaf/moonrise/patches/chunk_system/level/entity/EntityLookup.java
index 0d8ca784122c70d74ad49c20821b535e1cb967a9..b5b3c91aee7e143e6879a2216ef84d7e10c88399 100644
--- a/ca/spottedleaf/moonrise/patches/chunk_system/level/entity/EntityLookup.java
+++ b/ca/spottedleaf/moonrise/patches/chunk_system/level/entity/EntityLookup.java
@@ -51,6 +51,7 @@ public abstract class EntityLookup implements LevelEntityGetter<Entity> {
     protected final ConcurrentChainedLong2ReferenceHashTable<Entity> entityById = new ConcurrentChainedLong2ReferenceHashTable<>();
     protected final ConcurrentHashMap<UUID, Entity> entityByUUID = new ConcurrentHashMap<>();
     protected final EntityList accessibleEntities = new EntityList();
+    public final it.unimi.dsi.fastutil.objects.ReferenceArrayList<Entity> weakEntityList = new it.unimi.dsi.fastutil.objects.ReferenceArrayList<>(); // Leaf - Rewrite entity despawn time
 
     public EntityLookup(final Level world, final LevelCallback<Entity> worldCallback) {
         this.world = world;
@@ -264,6 +265,21 @@ public abstract class EntityLookup implements LevelEntityGetter<Entity> {
         try {
             final Boolean ticketBlockBefore = this.blockTicketUpdates();
             try {
+                // Leaf start - Rewrite entity despawn time
+                if (org.dreeam.leaf.config.modules.opt.DespawnTime.proactiveWeakLoading) {
+                    boolean isWeakNow = newVisibility == Visibility.TRACKED;
+                    boolean wasWeakBefore = oldVisibility == Visibility.TRACKED;
+
+                    if (entity.despawnTime >= 0) {
+                        if (isWeakNow && !wasWeakBefore) {
+                            this.weakEntityList.add(entity);
+                            entity.leaf$isWeakLoaded = true;
+                        } else if (!isWeakNow && wasWeakBefore) {
+                            entity.leaf$isWeakLoaded = false;
+                        }
+                    }
+                }
+                // Leaf end - Rewrite entity despawn time
                 ((ChunkSystemEntity)entity).moonrise$setUpdatingSectionStatus(true);
                 try {
                     if (created) {
@@ -504,6 +520,7 @@ public abstract class EntityLookup implements LevelEntityGetter<Entity> {
                 LOGGER.warn("Failed to remove entity " + entity + " from entity slices (" + sectionX + "," + sectionZ + ")");
             }
         }
+        if (org.dreeam.leaf.config.modules.opt.DespawnTime.proactiveWeakLoading) entity.leaf$isWeakLoaded = false; // Leaf - Rewrite entity despawn time
         ((ChunkSystemEntity)entity).moonrise$setSectionX(Integer.MIN_VALUE);
         ((ChunkSystemEntity)entity).moonrise$setSectionY(Integer.MIN_VALUE);
         ((ChunkSystemEntity)entity).moonrise$setSectionZ(Integer.MIN_VALUE);
@@ -942,6 +959,34 @@ public abstract class EntityLookup implements LevelEntityGetter<Entity> {
         }
     }
 
+    // Leaf start - Rewrite entity despawn time
+    public void leaf$tickWeakEntities() {
+        if (this.weakEntityList.isEmpty()) return;
+
+        for (int i = this.weakEntityList.size() - 1; i >= 0; i--) {
+            Entity entity = this.weakEntityList.get(i);
+
+            if (!entity.leaf$isWeakLoaded) {
+                removeAtIndex(this.weakEntityList, i);
+                continue;
+            }
+
+            if (entity.isExceedingDespawnTime(true)) {
+                entity.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.DESPAWN);
+                entity.leaf$isWeakLoaded = false;
+                removeAtIndex(this.weakEntityList, i);
+            }
+        }
+    }
+
+    private static void removeAtIndex(List<Entity> list, int index) {
+        int lastIndex = list.size() - 1;
+        if (index != lastIndex) {
+            list.set(index, list.get(lastIndex));
+        }
+        list.remove(lastIndex);
+    }
+    // Leaf end - Rewrite entity despawn time
     public static final class ChunkSlicesRegion {
 
         private final ChunkEntitySlices[] slices = new ChunkEntitySlices[REGION_SIZE * REGION_SIZE];
diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index 081918bec2e4a2c7dbf084a00b25e514bed8144b..1e02d238a6e75a97128fc3fd27cad9efd465f692 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -1136,6 +1136,7 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
         this.debugSynchronizers.tick(this.server.debugSubscribers());
         profiler.pop();
         this.chunkSource.randomState().garbageCollect();
+        if (org.dreeam.leaf.config.modules.opt.DespawnTime.proactiveWeakLoading) this.moonrise$getEntityLookup().leaf$tickWeakEntities(); // Leaf - Rewrite entity despawn time
         if (org.dreeam.leaf.config.modules.async.MultithreadedTracker.enabled) { this.leaf$asyncTracker.onEntitiesTickEnd(); } // Leaf - Multithreaded tracker
     }
 
diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index e8a791a33d6fb248dc2c542716f75cf5e424f66e..2adc004b78681b7289690d352ae132cdc938bd16 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -388,12 +388,14 @@ public abstract class Entity
     // Paper end
     public boolean freezeLocked = false; // Paper - Freeze Tick Lock API
     public boolean fixedPose = false; // Paper - Expand Pose API
-    private final int despawnTime; // Paper - entity despawn time limit
+    public final int despawnTime; // Paper - entity despawn time limit // Leaf - Rewrite entity despawn time - private -> public
     public int totalEntityAge; // Paper - age-like counter for all entities
+    private int lastTickTime; // Leaf - Rewrite entity despawn time
     public boolean activatedPriorityReset = false; // Pufferfish - DAB
     public int activatedPriority = org.dreeam.leaf.config.modules.opt.DynamicActivationofBrain.maximumActivationPrio; // Pufferfish - DAB (golf score)
     public final io.papermc.paper.entity.activation.ActivationType activationType = io.papermc.paper.entity.activation.ActivationType.activationTypeFor(this); // Paper - EAR 2/tracking ranges
     public @Nullable Boolean immuneToFire = null; // Purpur - Fire immune API
+    public boolean leaf$isWeakLoaded = false; // Leaf - Rewrite entity despawn time
     // Paper start - EAR 2
     public final boolean defaultActivationState;
     public long activatedTick = Integer.MIN_VALUE;
@@ -418,6 +420,7 @@ public abstract class Entity
     private int sectionY = Integer.MIN_VALUE;
     private int sectionZ = Integer.MIN_VALUE;
     private boolean updatingSectionStatus;
+    private final boolean shouldSkipBaseDespawnCheck = this instanceof net.minecraft.world.entity.projectile.ThrowableProjectile; // Leaf - Rewrite entity despawn time
 
     @Override
     public final boolean moonrise$isHardColliding() {
@@ -631,6 +634,7 @@ public abstract class Entity
         this.setPos(0.0, 0.0, 0.0);
         this.eyeHeight = this.dimensions.eyeHeight();
         this.despawnTime = level == null || type == EntityTypes.PLAYER ? -1 : level.paperConfig().entities.spawning.despawnTime.getOrDefault(type, io.papermc.paper.configuration.type.number.IntOr.Disabled.DISABLED).or(-1); // Paper - entity despawn time limit
+        this.updateLastTick(); // Leaf - Rewrite entity despawn time
     }
 
     public boolean isColliding(final BlockPos pos, final BlockState state) {
@@ -914,12 +918,19 @@ public abstract class Entity
     }
 
     public void tick() {
+        // Leaf start - Rewrite entity despawn time
+        if (!this.shouldSkipBaseDespawnCheck && this.detectDespawnTime()) {
+            return;
+        }
+        /*
         // Paper start - entity despawn time limit
         if (this.despawnTime >= 0 && this.totalEntityAge >= this.despawnTime) {
             this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.DESPAWN);
             return;
         }
         // Paper end - entity despawn time limit
+         */
+        // Leaf end - Rewrite entity despawn time
         this.baseTick();
     }
 
@@ -934,8 +945,55 @@ public abstract class Entity
         }
 
         this.tickCount++;
-        this.totalEntityAge++; // Paper - age-like counter for all entities
+        this.updateEntityAge(); // Paper - age-like counter for all entities // Leaf - Rewrite entity despawn time
+    }
+
+    // Leaf start - Rewrite entity despawn time
+    private int calculateMissedTicks() {
+        return net.minecraft.server.MinecraftServer.currentTick - this.lastTickTime;
+    }
+
+    public final boolean detectDespawnTime() {
+        return this.detectDespawnTime(false);
+    }
+
+    public final boolean detectDespawnTime(final boolean needCompensation) {
+        boolean exceeded = this.isExceedingDespawnTime(needCompensation);
+        if (exceeded) {
+            this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.DESPAWN);
+        }
+        return exceeded;
+    }
+
+    public final boolean isExceedingDespawnTime(boolean needCompensation) {
+        if (this.despawnTime >= 0) {
+            int compensatedAge = needCompensation ? this.getCompensatedEntityAge() : this.totalEntityAge;
+            return compensatedAge >= this.despawnTime;
+        }
+        return false;
+    }
+
+    private int getCompensatedEntityAge() {
+        int missedTicks = this.calculateMissedTicks();
+        if (missedTicks > 0) {
+            return this.totalEntityAge + missedTicks;
+        }
+        return this.totalEntityAge;
+    }
+
+    public void updateEntityAge() {
+        int missedTicks = this.calculateMissedTicks();
+        if (missedTicks > 1) {
+            this.totalEntityAge += missedTicks - 1;
+        }
+        this.totalEntityAge++;
+        this.updateLastTick();
+    }
+
+    protected void updateLastTick() {
+        this.lastTickTime = net.minecraft.server.MinecraftServer.currentTick;
     }
+    // Leaf end - Rewrite entity despawn time
 
     // CraftBukkit start
     public void postTick() {
@@ -2838,7 +2896,7 @@ public abstract class Entity
             // Leaf end - Fix Pufferfish and Purpur patches - Use effective default value for max air supply
                 output.putInt("Bukkit.MaxAirSupply", this.getMaxAirSupply());
             }
-            output.putInt("Spigot.ticksLived", this.totalEntityAge); // Paper
+            output.putInt("Spigot.ticksLived", this.getCompensatedEntityAge()); // Paper // Leaf - Rewrite entity despawn time
             // CraftBukkit end
             output.storeNullable("CustomName", ComponentSerialization.CODEC, this.getCustomName());
             if (this.isCustomNameVisible()) {
@@ -5573,7 +5631,8 @@ public abstract class Entity
     public boolean shouldBeSaved() {
         return (this.removalReason == null || this.removalReason.shouldSave())
             && !this.isPassenger()
-            && (!this.isVehicle() || !((ca.spottedleaf.moonrise.patches.chunk_system.entity.ChunkSystemEntity)this).moonrise$hasAnyPlayerPassengers()); // Paper - rewrite chunk system
+            && (!this.isVehicle() || !((ca.spottedleaf.moonrise.patches.chunk_system.entity.ChunkSystemEntity)this).moonrise$hasAnyPlayerPassengers())
+            && !this.isExceedingDespawnTime(true); // Paper - rewrite chunk system // Leaf - Rewrite entity despawn time
     }
 
     @Override
diff --git a/net/minecraft/world/entity/projectile/ThrowableProjectile.java b/net/minecraft/world/entity/projectile/ThrowableProjectile.java
index 82358c4f98a2720da6adf8c63616a810c2117919..aab9ef42324619dcef6afcba040e4a01b69822a9 100644
--- a/net/minecraft/world/entity/projectile/ThrowableProjectile.java
+++ b/net/minecraft/world/entity/projectile/ThrowableProjectile.java
@@ -42,6 +42,11 @@ public abstract class ThrowableProjectile extends Projectile {
 
     @Override
     public void tick() {
+        // Leaf start - Rewrite entity despawn time - Check earlier for projectiles
+        if (this.detectDespawnTime()) {
+            return;
+        }
+        // Leaf end - Rewrite entity despawn time - Check earlier for projectiles
         this.handleFirstTickBubbleColumn();
         this.applyGravity();
         this.applyInertia();
